pennylane/gradients/general_shift_rules.py

```diff
@@ -451,16 +402,26 @@ def generate_multishifted_tapes(tape, indices, shifts, multipliers=None):
             and ``multipliers``.
     """
     params = list(tape.get_parameters())
-    multipliers = np.ones_like(shifts) if multipliers is None else multipliers
+    shifts = qml.math.array(shifts)
+    if len(shifts.shape) == 1:
+        shifts = shifts.reshape((shifts.shape[0], 1))
+        indices = [indices]
+    if multipliers is None:
+        multipliers = np.ones_like(shifts)
+    else:
+        multipliers = qml.math.array(multipliers)
+        if len(multipliers.shape) == 1:
+            multipliers = multipliers.reshape((multipliers.shape[0], 1))

```

Comment:
@dwierichs out of curiosity, how does this perform compared to when there was a separate `generate_multishifted_tapes`? I admit, I find it easier to read the code of two separate functions, compared to one merged one.

Reply:
So kind of lean towards keeping two separate functions if there are no major impact either way? But I am not too strongly opinionated here.

Reply:
I _tend_ to prefer code duplication _iff_:

- The duplicated section is short
- It improves readability

But this is more personal preference, and of course it introduces the risk of duplicating bugs/issues.

Another approach could be to:

- Keep two separate public functions
- Factor out the shared logic into a private function

This is a 'compromise' in a sense between the two approaches.

What do you prefer?

> A case for this is maybe already made by the previous inconsistency between generate_shifted_tapes and generate_multishifted_tapes of whether shifted parameters were cast to the original parameter dtype or not.

Yes, this is a good argument for having a single public function!

Reply:
Oh there is also a third approach: single public function, private implementations:

```python
def generate_shifted_tapes(tape, idx, shifts, multipliers=None):
    if multipliers:
        return _generate_multishifted_tapes(...)
    return _generate_shifted_tapes(...)
```

Maybe this is the best way 🤔 This way you (a) preserve readability, while also ensuring both functions conform to the same signature.

Reply:
I'm wondering: For these two functions, there is a common core functionality, which is to create a new, modified tape. 
The functions would more likely differ in the expected input formats, I suppose. So maybe two public functions with a shared private "core" function would be a suitable way to go?

Another option is to reduce the logic of the new `generate_shifted_tapes` so that it only ever accepts a `Sequence` for `indices`, and a `Sequence` of shift value `Sequence`s as it is needed for multishifted tapes. Then we could require developers to use it like `generate_shifted_tapes(tape, [idx], [shifts], [multipliers])` in the case of a single shifted parameter.

Reply:
>  So maybe two public functions with a shared private "core" function would be a suitable way to go?

I'm happy with this, but not too opinionated 🙂 I favour this over your second option only because I imagine it will be easier to describe usage in the docstrings

Reply:
I ended up splitting them entirely again, so basically reverted the change.
I took another detailed look at [the PR](https://github.com/PennyLaneAI/pennylane/pull/2120) introducing the casting and it makes sense to leave the inconsistency regarding data type casting in, because it is required when using the gradients module in the context of one of the frameworks, and for vjps, it is implemented in a separate place, whereas `param_shift_hessian` still has this somewhat hybrid character of both a low- and high-level function.